mllib/src/main/scala/org/apache/spark/mllib/regression/StreamingLinearRegressionWithSGD.scala

```diff
@@ -47,6 +52,7 @@ class StreamingLinearRegressionWithSGD private[mllib] (
     private var numIterations: Int,
     private var miniBatchFraction: Double)
   extends StreamingLinearAlgorithm[LinearRegressionModel, LinearRegressionWithSGD]
+  with StreamingDecaySetter[StreamingLinearRegressionWithSGD]
```

Comment:
OK, I think I get why you have both `StreamingDecay` and `StreamingDecaySetter`; it's because `getDiscount` is also needed in `trainOn` but you also need to get the concrete type `StreamingLinearRegressionWithSGD` for returning from the setters (e.g. `setHalfLife`), right?

I'm still for having a single trait over trait inheritance, especially if the reason we have this inheritance is to get the concrete type and we won't be adding other concrete `StreamingDecay` implementations soon.

We could just have a single `private[mllib] trait StreamingDecay` and have `StreamingLinearAlgorithm extends StreamingDecay[A]`; the concrete type should be available in `A` and return properly. Let me know if that works...